pkg/analysis/markerscope/analyzer.go

```diff
@@ -0,0 +1,491 @@
+	"slices"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+
+	kalerrors "sigs.k8s.io/kube-api-linter/pkg/analysis/errors"
+	markershelper "sigs.k8s.io/kube-api-linter/pkg/analysis/helpers/markers"
+	"sigs.k8s.io/kube-api-linter/pkg/analysis/utils"
+)
+
+const (
+	name = "markerscope"
+)
+
+func init() {
+	// Register all markers we want to validate scope for
+	defaults := DefaultMarkerRules()
+	markers := make([]string, 0, len(defaults))
+
+	for marker := range defaults {
+		markers = append(markers, marker)
+	}
+
+	markershelper.DefaultRegistry().Register(markers...)
+}
+
+type analyzer struct {
+	markerRules         map[string]MarkerScopeRule
+	policy              MarkerScopePolicy
+	allowDangerousTypes bool
+}
+
+// newAnalyzer creates a new analyzer.
+func newAnalyzer(cfg *MarkerScopeConfig) *analysis.Analyzer {
+	if cfg == nil {
+		cfg = &MarkerScopeConfig{}
+	}
+
+	a := &analyzer{
+		markerRules:         mergeMarkerRules(DefaultMarkerRules(), cfg.MarkerRules),
+		policy:              cfg.Policy,
+		allowDangerousTypes: cfg.AllowDangerousTypes,
+	}
+
+	// Register all markers (both default and custom) with the markers helper
+	// This must be done before the analyzer runs because the markers helper
+	// analyzer needs to know about these markers
+	for marker := range a.markerRules {
+		markershelper.DefaultRegistry().Register(marker)
+	}
+
+	// Set default policy if not specified
+	if a.policy == "" {
+		a.policy = MarkerScopePolicyWarn
+	}
+
+	return &analysis.Analyzer{
+		Name:             name,
+		Doc:              "Validates that markers are applied in the correct scope.",
+		Run:              a.run,
+		Requires:         []*analysis.Analyzer{inspect.Analyzer, markershelper.Analyzer},
+		RunDespiteErrors: true,
+	}
+}
+
+// mergeMarkerRules merges custom marker rules with default marker rules.
+// Custom rules take precedence over default rules for the same marker.
+func mergeMarkerRules(defaults, custom map[string]MarkerScopeRule) map[string]MarkerScopeRule {
+	merged := make(map[string]MarkerScopeRule, len(defaults)+len(custom))
+
+	// Copy all default rules
+	maps.Copy(merged, defaults)
+
+	// Override with custom rules
+	maps.Copy(merged, custom)
+
+	return merged
+}
+
+func (a *analyzer) run(pass *analysis.Pass) (any, error) {
+	inspect, ok := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	if !ok {
+		return nil, kalerrors.ErrCouldNotGetInspector
+	}
+
+	markersAccess, ok := pass.ResultOf[markershelper.Analyzer].(markershelper.Markers)
+	if !ok {
+		return nil, kalerrors.ErrCouldNotGetMarkers
+	}
+
+	// Check field markers and type markers
+	nodeFilter := []ast.Node{
+		(*ast.Field)(nil),
+		(*ast.GenDecl)(nil),
+	}
+
+	inspect.Preorder(nodeFilter, func(n ast.Node) {
+		switch node := n.(type) {
+		case *ast.Field:
+			a.checkFieldMarkers(pass, node, markersAccess)
+		case *ast.GenDecl:
+			a.checkTypeMarkers(pass, node, markersAccess)
+		}
+	})
```

Comment:
We have our own inspector helper that we prefer to use rather than this built in one, please check other linters and copy that pattern